PMax Assets - Fetch assets groups

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

This PR adds the functionality to fetch the asset groups for a specific campaign ID. The asset group will contain assets of the following type (when present):

Asset Field Type|
--|
HEADLINE|
LONG_HEADLINE|
DESCRIPTION|
BUSINESS_NAME|
MARKETING_IMAGE|
SQUARE_MARKETING_IMAGE|
LOGO|
CALL_TO_ACTION_SELECTION|




Example response:

```
 [
 {
        "id": 6443481181,
        "final_url": "https://mystagingwebsite.com",
        "display_url_path": [
            "",
            ""
        ],
        "assets": {
            "square_marketing_image": [
                {
                    "id": 23933203451,
                    "content": "https://tpc.googlesyndication.com/simgad/184308248126859863561"
                },
                {
                    "id": 23946157339,
                    "content": "https://tpc.googlesyndication.com/simgad/70392509749665497401"
                }
            ],
            "marketing_image": [
                {
                    "id": 23933203454,
                    "content": "https://tpc.googlesyndication.com/simgad/113463390769957791581"
                }
            ],
         .....more assets
 }
]
```

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. As we are using some new imports from the GoogleAds Library, you will need to remove your vendor folder `rm -rf vendor/*` and install again the dependencies `composer install`
2. Go to your Ads Account, choose one campaign and get the campaignID. The campaign ID can be found in the URL with the param `campaignId`
3. Go to the Asset Group Sections and choose an asset group.

![image](https://user-images.githubusercontent.com/2488994/208157284-311e65da-b2f6-42b5-afea-f438625545a6.png)

5. Make a request to `GET ads/campaigns/CAMPAIGN_ID/assetgroups`
6. Compare the API results with the Asset Group in your Ads Account.
7. If you do not have assets, add some of them and compare the results. By default the asset group is empty when is created using the GL&A Plugin.

 
### Additional details:
 - For the MVP, the number of asset groups returned is limited to 1.
 - If there are multiple asset groups the Ads API returns by default the oldest Asset Group.
<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

jorgemd24 committed:
Add endpoint assetgroup controller

jorgemd24 committed:
Add method get_asset_groups

jorgemd24 committed:
get asset groups for a specific campaign

jorgemd24 committed:
Add ads assetgroupasset query

jorgemd24 committed:
Add function get_asset_groups_assets

jorgemd24 committed:
Add fields : call to action, path1 and path2

jorgemd24 committed:
Split logic in different classes

jorgemd24 committed:
Add tests AssetGroupAssets

jorgemd24 committed:
Add tests AssetGroup

jorgemd24 committed:
Add exeption asset group controller

jorgemd24 committed:
Add vars defs

jorgemd24 committed:
add where clausures to get_asset_group_assets query

jorgemd24 committed:
Fix phpunit tests

jorgemd24 committed:
Add test case if not asset group ids

jorgemd24 committed:
Add AssetGroupController schema

jorgemd24 committed:
Adjust final url field

jorgemd24 committed:
Add AssetGroupController test

jorgemd24 committed:
Add default case

jorgemd24 committed:
Adjust AdsAssetGroupAssetQuery

jorgemd24 committed:
Add pageSize

jorgemd24 committed:
Fix asset group tests

jorgemd24 committed:
Add get_asset_field_types

jorgemd24 committed:
Update phpdocs

jorgemd24 committed:
Rename function

jorgemd24 committed:
add AssetFieldType

jorgemd24 committed:
Add AssetFieldTypes

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -0,0 +1,204 @@
+	 */
+	public function register_routes(): void {
+		$this->register_route(
+			'ads/campaigns/(?P<id>[\d]+)/assetgroups',
```
Some naming/location about this controller. Probably I'm missing concepts. But just to be sure :) 
 
❓  Why this file is not under Controller/Assets as `AssetSuggestionsController`
❓  I would name it under `assets/(?P<id>[\d]+)/groups` 

💅  In any case, instead of `assetgroups` I'd suggest `asset-groups`

jorgemd24 replied:
> ❓ Why this file is not under Controller/Assets as AssetSuggestionsController

I have some doubts about the location too, my thinking was Asset Groups are linked to a campaign, campaign are in the `Ads` directory, and asset groups belongs as well to the "Ads" section, and I was thinking as well about moving the `AssetSuggestionsController ` to the `Ads` directory. WDYT? Or do you think it is better to leave a specific folder for the `Assets`?

> ❓ I would name it under assets/(?P<id>[\d]+)/groups

Not sure about this one, because it seems that groups are under `assets` but the hierarchy is more like this:

Campaign -> Asset Group -> Asset Group Asset (the link between Asset Group and Asset) -> Assets. Also, the terminology in the Ads API is `Asset Group`.

> 💅 In any case, instead of assetgroups I'd suggest asset-groups

👍

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -0,0 +1,204 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class AssetGroupController
```
💅  Instead of adding the trivial `Class AssetGroupController` I'd suggest to add some explanation about the responsibility of this controller.

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -0,0 +1,204 @@
+							'items'    => $this->get_schema_asset(),
+							'required' => false,
+						],
+						'call_to_action_selection' => [
+							'type'     => 'array',
+							'items'    => $this->get_schema_asset(),
+							'required' => false,
+						],
```
💅  What about wrap all of these into:

```php
... 
'long_headline'  => $this->get_schema_asset_group(),
'call_to_action_selection' => $this->get_schema_asset_group(),
... 

protected function get_schema_asset_group() {
 return [
   'type'     => 'array',
   'items'    => $this->get_schema_asset(), 
   'required' => false,
 ]
}



```

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -0,0 +1,204 @@
+	protected function get_asset_groups_by_campaign_id_callback(): callable {
+		return function( Request $request ) {
+			try {
+				$campaign_id = $request['id'];
```
💅  Ideally we can use `$request->get_param('id);`

puntope commented on src/API/Google/AdsAssetGroup.php:
```diff
@@ -190,11 +205,80 @@ protected function listing_group_delete_operations(): array {
+
+		/** @var GoogleAdsRow $row */
+		foreach ( $asset_group_results->getPage()->getIterator() as $row ) {
+			$asset_groups_converted[ $row->getAssetGroup()->getId() ] = $this->convert_asset_group( $row );
```
❓  Why we are saving the assets in `$asset_groups_converted` as an associative array if we return only the values ?

jorgemd24 replied:
I think it is handy to leave the `id` as a `key` because it allows retrieving the `ids` easily using `array_keys( $asset_groups );`.  For example in this case: https://github.com/woocommerce/google-listings-and-ads/blob/2ac0403841d865a5530546d8c87cdccbc364d8bf/src/API/Google/AdsAssetGroup.php#L258

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -0,0 +1,103 @@
+
+
+/**
+ * Class AdsAssetGroupAsset
```
💅  Same as other files. It would be cool to add some brief explanation about the file.

puntope commented on src/API/Google/AssetFieldType.php:
```diff
@@ -0,0 +1,153 @@
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\API\Google
+ */
+class AssetFieldType extends StatusMapping {
```
💅  WDYT about renaming this to `AssetFieldTypeMapping`?

jorgemd24 replied:
The other classes that are implementing `StatusMapping` do not use the suffix Mapping. See: 

- https://github.com/woocommerce/google-listings-and-ads/blob/0ffb057cbe73a3b1989bb35babed13899da1c4cf/src/API/Google/CampaignType.php#L17
- https://github.com/woocommerce/google-listings-and-ads/blob/0ffb057cbe73a3b1989bb35babed13899da1c4cf/src/API/Google/BillingSetupStatus.php#L15

So I think it is ok if we leave it as it is.

puntope commented on src/API/Google/AdsAssetGroup.php:
```diff
@@ -190,11 +205,80 @@ protected function listing_group_delete_operations(): array {
+	 *
+	 * @return array The asset groups for the campaign.
+	 */
+	public function get_asset_groups_by_campaign_id( int $campaign_id, bool $include_assets = true ): array {
```
❓  In regards `$include_assets` param. Is there any case we don't want to get the assets?

jorgemd24 replied:
Currently, we are showing by default only one asset group including the assets however in the future if we decide to work with multiple asset groups maybe could be necessary to list all the asset groups linked to a campaign and later load the assets for a specific asset group. 

Also, I added the param because it was not adding complexity to the function and was making the function more flexible.

puntope commented on src/API/Google/Query/AdsAssetGroupAssetQuery.php:
```diff
@@ -0,0 +1,22 @@
+	/**
+	 * Query constructor.
+	 */
+	public function __construct() {
```
💅  Would be cool to ad some ref about the parameters available here in case another dev needs to update this query.

jorgemd24 replied:
Do you mean the methods and properties available from the `AdsQuery` and `Query` class?

puntope replied:
I was thinking about a link where you found the documentation for the available columns and params in the query.

puntope commented:
Howdy @jorgemd24  I tested this PR. Thanks for your work here.

✅  I checked to get the assets of an existen campaign with no asset inside the asset group and it gave me this result;

```json
[
    {
        "id": 6445482928,
        "final_url": "",
        "display_url_path": [
            "",
            ""
        ],
        "assets": []
    }
]
```

✅  I checked to get the assets of an existen campaign with a single asset group and it gave me the right result.
✅  I checked to get the assets of an existen campaign with a two asset groups and it gave me the oldest asset group result.
✅  I checked to get a result of a non existent campaign and it gave me empty array []

💅  / ❓  I left some comments in regards the code

jorgemd24 committed:
use get_param method to get the param

jorgemd24 committed:
Add phpdoc for AssetGroupController

jorgemd24 committed:
Add schema schema_field_type_asset

jorgemd24 committed:
Add phpdoc for class AdsAssetGroupAsset

jorgemd24 committed:
Update controller route asset-groups

jorgemd24 committed:
Add ApiException try/catch

jorgemd24 committed:
Add try/catch for asset group assets

jorgemd24 commented:
Thanks, @puntope for reviewing this!

I addressed your comments and replied to your questions. I also realized that we could add a try/catch in the query requests, so I went ahead and I added it.

puntope reviewed (approved):
✅  Approved

Thanks for the adjustments and answering my questions 👍

jorgemd24 committed:
Move AssetSuggestions from Assets folder to Ads folder
